044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,154 @@
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is hard-coded to 9090.
+If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
+* The current-state metric could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
+* The broker nodes will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
```

Comment:
A sentence here about this `readiness` not blocking connections from other controllers would really help clarify, I think.